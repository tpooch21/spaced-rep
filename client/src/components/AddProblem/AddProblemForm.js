import {
  Box,
  Input,
  Radio,
  RadioGroup,
  Stack,
  Flex,
  FormControl,
  FormErrorMessage,
  Button,
} from "@chakra-ui/react";

// Misc
import { useForm } from "react-hook-form";
import axios from "axios";
import { isHttpUri, isHttpsUri } from "valid-url";

/**
 * Name
 * URL
 * Number (optional)
 * Difficulty (optional)
 */
const url = "http://localhost:3001";

const AddProblemForm = ({ add }) => {
  const { handleSubmit, register, errors, reset } = useForm();
  const onSubmit = (data) => {
    const { problemName, problemURL, problemId, difficulty } = data;

    axios
      .post(url, {
        name: problemName,
        url: problemURL,
        id: problemId,
        difficulty: difficulty,
      })
      .then((res) => {
        console.log(res.data);
        add(res.data);
      })
      .catch((err) => console.error(err));

    reset();
  };

  const validateURL = (value) => {
    if (isHttpsUri(value) || isHttpUri(value)) {
      return true;
    }
    return "URL is invalid";
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Flex
        w={80}
        p={4}
        mt={4}
        mb={3}
        bg="gray.300"
        rounded="sm"
        fontFamily="main"
        color="green.200"
        direction="column"
        align="center"
      >
        <FormControl isInvalid={errors.problemName}>
          <Input
            name="problemName"
            placeholder="*Name"
            size="sm"
            bg="gray.600"
            fontWeight="700"
            ref={register({ required: "Name is required" })}
            data-testid="name"
          />
          <FormErrorMessage fontSize="xs" mt={1} data-testid="name-msg">
            {errors.problemName && errors.problemName.message}
          </FormErrorMessage>
        </FormControl>
        <FormControl isInvalid={errors.problemURL}>
          <Input
            name="problemURL"
            placeholder="*URL"
            size="sm"
            mt={3}
            bg="gray.600"
            border="none"
            fontWeight="700"
            ref={register({
              required: "URL is required",
              validate: validateURL,
            })}
            data-testid="url"
          />
          <FormErrorMessage fontSize="xs" mt={1} data-testid="url-msg">
            {errors.problemURL && errors.problemURL.message}
          </FormErrorMessage>
        </FormControl>
        <Flex mt={3} spacing={3} align="center" justify="space-between" w={72}>
          <FormControl>
            <Input
              name="problemId"
              placeholder="ID"
              size="sm"
              bg="gray.600"
              border="none"
              width={14}
              fontWeight="700"
              ref={register}
            />
          </FormControl>
          <FormControl>
            <RadioGroup color="gray.600" fontWeight="500" fontSize="sm" pr={2}>
              <Stack direction="row">
                <Radio
                  name="difficulty"
                  size="sm"
                  value="Easy"
                  colorScheme="green"
                  ref={register}
                >
                  Easy
                </Radio>
                <Radio
                  name="difficulty"
                  size="sm"
                  value="Medium"
                  colorScheme="yellow"
                  ref={register}
                >
                  Medium
                </Radio>
                <Radio
                  name="difficulty"
                  size="sm"
                  value="Hard"
                  colorScheme="red"
                  ref={register}
                >
                  Hard
                </Radio>
              </Stack>
            </RadioGroup>
          </FormControl>
        </Flex>
        <Button
          type="submit"
          mt={3}
          p={2}
          bg="green.200"
          color="gray.600"
          w={36}
        >
          Submit
        </Button>
      </Flex>
      <Box
        w={80}
        h={7}
        mt={0}
        borderRadius="sm"
        bg={errors.problemName || errors.problemURL ? "red.300" : "green.200"}
      ></Box>
    </form>
  );
};

export default AddProblemForm;
