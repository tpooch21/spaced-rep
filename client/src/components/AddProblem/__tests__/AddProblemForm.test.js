/**
 * Only submit form when information is valid
 * Essentially testing the validation of react-hook-form
 */
import AddProblemForm from "../AddProblemForm";
import renderer from "react-test-renderer";
import { render, screen, fireEvent, act } from "@testing-library/react";

const mockSubmitData = jest.fn(() => {
  return Promise.resolve("Success");
});

it("displays a form with no error messages on initial render", () => {
  const tree = renderer.create(<AddProblemForm />);
  expect(tree).toMatchSnapshot();
});

it("displays correct error message when name field is left blank", async () => {
  const { getByTestId, getByRole } = render(<AddProblemForm />);

  await act(async () => {
    fireEvent.submit(getByRole("button"));
  });

  expect(getByTestId("name-msg").innerHTML).toEqual("Name is required");
});

it("displays correct error message when url field is left blank", async () => {
  const { getByTestId, getByRole } = render(<AddProblemForm />);

  await act(async () => {
    fireEvent.submit(getByRole("button"));
  });

  expect(getByTestId("url-msg").innerHTML).toEqual("URL is required");
});

it("displays correct error message when url is invalid", async () => {
  const { getByTestId, getByRole } = render(<AddProblemForm />);

  await act(async () => {
    fireEvent.change(getByTestId("url"), {
      target: {
        value: "Invalid url",
      },
    });
    fireEvent.submit(getByRole("button"));
  });

  expect(getByTestId("url-msg").innerHTML).toEqual("URL is invalid");
});

// "does not submit data to backend when there are any errors"
