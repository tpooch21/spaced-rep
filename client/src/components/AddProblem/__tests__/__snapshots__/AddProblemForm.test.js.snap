// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a form with no error messages on initial render 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="css-107yhbn"
  >
    <div
      className="chakra-form-control css-1kxonj9"
      role="group"
    >
      <input
        aria-describedby="field-1-feedback"
        className="chakra-input css-at91oz"
        data-testid="name"
        disabled={false}
        id="field-1"
        name="problemName"
        onBlur={[Function]}
        onFocus={[Function]}
        placeholder="*Name"
        readOnly={false}
        required={false}
      />
    </div>
    <div
      className="chakra-form-control css-1kxonj9"
      role="group"
    >
      <input
        aria-describedby="field-2-feedback"
        className="chakra-input css-quw5d"
        data-testid="url"
        disabled={false}
        id="field-2"
        name="problemURL"
        onBlur={[Function]}
        onFocus={[Function]}
        placeholder="*URL"
        readOnly={false}
        required={false}
      />
    </div>
    <div
      className="css-11nn6z7"
      spacing={3}
    >
      <div
        className="chakra-form-control css-1kxonj9"
        role="group"
      >
        <input
          className="chakra-input css-1uxp5q"
          disabled={false}
          id="field-3"
          name="problemId"
          onBlur={[Function]}
          onFocus={[Function]}
          placeholder="ID"
          readOnly={false}
          required={false}
        />
      </div>
      <div
        className="chakra-form-control css-1kxonj9"
        role="group"
      >
        <div
          className="chakra-radio-group css-wotfcf"
          role="radiogroup"
        >
          <div
            className="chakra-stack css-101iv94"
          >
            <label
              className="chakra-radio css-78joka"
            >
              <input
                checked={false}
                className="chakra-radio__input"
                disabled={false}
                id="field-5"
                name="difficulty"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                readOnly={false}
                required={false}
                style={
                  Object {
                    "border": "0px",
                    "clip": "rect(0px, 0px, 0px, 0px)",
                    "height": "1px",
                    "margin": "-1px",
                    "overflow": "hidden",
                    "padding": "0px",
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "1px",
                  }
                }
                type="radio"
                value="easy"
              />
              <div
                aria-hidden={true}
                className="chakra-radio__control css-ssalds"
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
              />
              <div
                className="chakra-radio__label css-1bpjuuo"
                onMouseDown={[Function]}
                onTouchStart={[Function]}
              >
                Easy
              </div>
            </label>
            <label
              className="chakra-radio css-78joka"
            >
              <input
                checked={false}
                className="chakra-radio__input"
                disabled={false}
                id="field-5"
                name="difficulty"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                readOnly={false}
                required={false}
                style={
                  Object {
                    "border": "0px",
                    "clip": "rect(0px, 0px, 0px, 0px)",
                    "height": "1px",
                    "margin": "-1px",
                    "overflow": "hidden",
                    "padding": "0px",
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "1px",
                  }
                }
                type="radio"
                value="medium"
              />
              <div
                aria-hidden={true}
                className="chakra-radio__control css-ssalds"
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
              />
              <div
                className="chakra-radio__label css-1bpjuuo"
                onMouseDown={[Function]}
                onTouchStart={[Function]}
              >
                Medium
              </div>
            </label>
            <label
              className="chakra-radio css-78joka"
            >
              <input
                checked={false}
                className="chakra-radio__input"
                disabled={false}
                id="field-5"
                name="difficulty"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                readOnly={false}
                required={false}
                style={
                  Object {
                    "border": "0px",
                    "clip": "rect(0px, 0px, 0px, 0px)",
                    "height": "1px",
                    "margin": "-1px",
                    "overflow": "hidden",
                    "padding": "0px",
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "1px",
                  }
                }
                type="radio"
                value="hard"
              />
              <div
                aria-hidden={true}
                className="chakra-radio__control css-ssalds"
                onMouseDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
              />
              <div
                className="chakra-radio__label css-1bpjuuo"
                onMouseDown={[Function]}
                onTouchStart={[Function]}
              >
                Hard
              </div>
            </label>
          </div>
        </div>
      </div>
    </div>
    <button
      className="chakra-button css-1lf5o8h"
      type="submit"
    >
      Submit
    </button>
  </div>
  <div
    className="css-1eren5u"
  />
</form>
`;
